src/expr/src/interpret.rs

```diff
@@ -0,0 +1,1180 @@
+    Maybe,
+    /// We don't know any useful properties of this function.
+    No,
+}
+
+fn unary_monotonic(func: &UnaryFunc) -> Monotonic {
+    use crate::func::impls as funcs;
+    use Monotonic::*;
+    use UnaryFunc::*;
+    match func {
+        // Casts are generally monotonic.
+        CastUint64ToNumeric(_)
+        | CastNumericToUint64(_)
+        | CastNumericToMzTimestamp(_)
+        | CastUint64ToMzTimestamp(_)
+        | CastTimestampToMzTimestamp(_)
+        | CastTimestampTzToMzTimestamp(_) => Yes,
+        // Including JSON casts.
+        CastJsonbToNumeric(_)
+        | CastJsonbToBool(_)
+        | CastJsonbToString(_)
+        | CastJsonbToInt64(_)
+        | CastJsonbToFloat64(_) => Yes,
+
+        // Extracting the "most significant bits" of the a timestamp is monotonic.
+        ExtractTimestamp(funcs::ExtractTimestamp(units))
+        | ExtractTimestampTz(funcs::ExtractTimestampTz(units)) => match units {
+            DateTimeUnits::Epoch
+            | DateTimeUnits::Millennium
+            | DateTimeUnits::Century
+            | DateTimeUnits::Decade
+            | DateTimeUnits::Year => Yes,
+            _ => No,
+        },
+        DateTruncTimestamp(_) | DateTruncTimestampTz(_) => Yes,
+        // Negation is monotonically decreasing, but that's fine.
+        NegInt64(_) | NegNumeric(_) | Not(_) => Yes,
+        // Trivially monotonic, since all non-null values map to `false`.
+        IsNull(_) => Yes,
+        AbsInt64(_) => No,
+        _ => Maybe,
+    }
+}
+
+fn binary_monotonic(func: &BinaryFunc) -> (Monotonic, Monotonic) {
+    use BinaryFunc::*;
+    use Monotonic::*;
+
+    match func {
+        AddInt64 | MulInt64 | AddNumeric | MulNumeric | SubInt64 | SubNumeric => (Yes, Yes),
+        AddInterval | SubInterval => (Yes, Yes),
+        AddTimestampInterval
+        | AddTimestampTzInterval
+        | SubTimestampInterval
+        | SubTimestampTzInterval => (Yes, Yes),
+        // Monotonic in the left argument, but the right hand side has a discontinuity.
+        // Could be treated as monotonic if we also captured the valid domain of the function args.
+        DivInt64 | DivNumeric => (Yes, No),
+        Lt | Lte | Gt | Gte => (Yes, Yes),
+        _ => (Maybe, Maybe),
+    }
+}
+
+fn variadic_monotonic(func: &VariadicFunc) -> Monotonic {
+    use Monotonic::*;
+    use VariadicFunc::*;
+    match func {
+        And | Or => Yes,
+        _ => Maybe,
+    }
+}
+
+/// An inclusive range of non-null datum values.
+#[derive(Clone, Eq, PartialEq, Debug)]
+enum Values<'a> {
+    /// This range contains no values.
+    Empty,
+    /// An inclusive range. Invariant: the first element is always <= the second.
+    Within(Datum<'a>, Datum<'a>),
+    /// Constraints on structured fields, useful for recursive structures like maps.
+    /// Fields that are not present in the map default to Values::All.
+    Nested(BTreeMap<Datum<'a>, ResultSpec<'a>>),
```

Comment:
I see `Nested` for maps, but what is done for lists/arrays?

Reply:
A good question!

Currently they are treated like non-collection datums - ie. we'll keep the min and max list in the `Within` variant. This is well-defined but not very helpful.

A more useful approach would be to use the `Nested` variant for lists as well. In that case the keys in the map would be indexes into the list.

This is a slightly moot point at the moment for our desired Persist-stats use-case, since we do not collect statistics for lists or arrays. However, I'd be happy to switch to the second / more principled approach if that seems preferable.

Reply:
Makes sense! I don't have a great sense of what people are doing with lists, and that seems perfectly sensible for now.  If we, say, tracked list length we might be able to notice certain errors (or lack of them) ahead of time.

Adding a variant later should mean changing `Values` later and adding some cases for appropriate functions, which doesn't seem like a big change.